release-21.2: Limit concurrency of catchup scans.

miretskiy opened the pull request:
Backport:
  * 2/2 commits from "changefeedccl, kv: Add few metrics" (#77711)
  * 1/1 commits from " kv: Limit concurrency of catchup scans." (#77725)

Please see individual PRs for details.

/cc @cockroachdb/release

Release Justification: Stability improvement

Commit:
changefeedccl: Add a histogram metric on message sizes.

Add a histogram metric keeping track fo emitted messages sizes.

Release Note (enterprise change): Changefeed records message size
histogram.

Release Justification: Low impact change.

Commit:
kv: Add rangfeed metrics to dist sender.

Add 2 metrics to keep track of the total number of ranges
executing rangefeed, as well as the number of ranges currently
performing the catchup scan.

Release Notes: None
Release Justification: Low impact, observability improvement.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77932)
<!-- Reviewable:end -->

miretskiy commented:
@erikgrinaker fyi: had to push a fix to tolerate rangefeed error when catchpu reservation is non-nil.
Also, this release uses older Go -- which lacked math.MaxInt constant (which I use) -- switched to math.MaxInt64.

Commit:
kv: Limit concurrency of catchup scans.

Limit concurrency of the catchup scans.
Prior to this change, distsender would issue all of the rangefeed
calls immediately, and the the KV server deal with limiting
the number of concurrent catchup scans it can handle.  This has
a potential of using up entirety of the KV server budget by a single
rangefeed client (e.g. changefeed), thus starving all other users.

This PR adds a `kv.rangefeed.catchup_scan_concurrency` setting
to limit the number of catchup scans started by rangefeed client.

Release Notes (enterprise): Limit the number of concurent catchup scan
requests issued by rangefeed client.

Release Justfication: Important stability and scalability improvement
to prevent a single rangefeed client from consuming entirety of catchup
scan iterator budget on a KV server.

miretskiy commented:
Informs #77702
